File: misc_sorting_algos/MergeSort/merge_sort.py
'''
* Author: Tomoya Tokunaga(mailto: ttokunag@ucsd.edu)
*
'''

def merge_sort(arr):
    # an array length of <= 1 is already sorted
    if len(arr) <= 1:
        return arr
    # set a pivot to divide the given array into two
    pivot = len(arr)//2
    left_arr = merge_sort(arr[:pivot])
    right_arr = merge_sort(arr[pivot:])
    
    '''
    * the following is a helper method to recursively sort subarrayslef
    * @param left_array: the left divided array
    * @param right_array: the right divided array
    '''
    def merge(left_arr, right_arr):
        left_cur = right_cur = 0
        res = []
        # place a smaller element of two arrays to a smaller index
        while left_cur < len(left_arr) and right_cur < len(right_arr):
            if left_arr[left_cur] < right_arr[right_cur]:
                res.append(left_arr[left_cur])
                left_cur += 1
            else:
                res.append(right_arr[right_cur])
                right_cur += 1
        # once one array gets empty, append the other
        res.extend(left_arr[left_cur:])
        res.extend(right_arr[right_cur:])
        return res
    
    return merge(left_arr, right_arr)

File: Sorting_algos/misc_sorting_algos/MergeSort/MergeSort.java
package com.cse.ds.sort;

import com.cse.ds.Comparator;
import com.cse.ds.Sorting;

/**
 * 
 * @author Tomoya Tokunaga (mailto: ttokunag@ucsd.edu)
 *
 * This file implements Merge sort algorithm
 */
public class MergeSort<T> implements Sorting<T> {

	private Comparator<T> comparator;

	/*
    * MergeSort class requires one parameter, Comparator<T>
    */
	public MergeSort(Comparator<T> comparator) {
		this.comparator = comparator;
	}

	/*
    * description: sorts an array based on data type
	* @param: array, an array to be sorted
	* @param: ascending, true if an array to be sorted in the ascending order
    * return: void
    */
	@Override
	public void sort(T[] array, boolean ascending) {
		if (array.length > 1) {
			int mid = array.length / 2;
			Object[] left = new Object[mid];
			Object[] right = new Object[array.length - mid];
			for (int i = 0; i < mid; i++) {
				left[i] = array[i];
			}
			for (int i = 0; i < array.length - mid; i++) {
				right[i] = array[i + mid];
			}
			
			this.sort((T[])left, ascending);
			this.sort((T[])right, ascending);
			int i = 0;
			int j = 0;
			int k = 0;
			
			while (i < left.length && j < right.length) {
				if (this.comparator.comparison((T)left[i], (T)right[j], ascending)) {
					array[k] = (T)left[i];
					i++;
				}
				else {
					array[k] = (T)right[j];
					j++;
				}
				k++;
			}
			
			while (i < left.length) {
				array[k] = (T)left[i];
				i++;
				k++;
			}
			
			while (j < right.length) {
				array[k] = (T)right[j];
				j++;
				k++;
			}
		}
	}
}
